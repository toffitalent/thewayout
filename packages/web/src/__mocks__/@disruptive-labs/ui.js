const actual = jest.requireActual('@disruptive-labs/ui');

const showToast = jest.fn();

actual.Toast.show = showToast;
actual.Toast.danger = showToast;
actual.Toast.info = showToast;
actual.Toast.success = showToast;
actual.Toast.warning = showToast;

module.exports = {
  ...actual,
  showToast,
};
