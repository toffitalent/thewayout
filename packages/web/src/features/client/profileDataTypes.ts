import {
  FieldComponentProps,
  FieldType,
  WizardChoicesProps,
  WizardStep,
} from '@disruptive-labs/ui';
import {
  Age,
  Education,
  Ethnicity,
  ExperienceSkills,
  Gender,
  JusticeStatus,
  Languages,
  License,
  MaritalStatus,
  OffenseWizardType,
  Orientation,
  PersonalStrengths,
  ReferredBy,
  RelativeExperience,
  ReleasedAt,
  Religion,
  State,
  StateFederal,
  Support,
  TimeServed,
  VeteranBranchOfService,
  VeteranCampaign,
  VeteranOrJustice,
  VeteranRank,
  VeteranStatus,
  VeteranTypeDischarge,
} from '@two/shared';

export enum StepsClient {
  veteranOrJustice = 'veteranOrJustice',

  veteranService = 'veteranService',
  veteranReservist = 'veteranReservist',
  veteranCampaigns = 'veteranCampaigns',
  veteranDates = 'veteranDates',
  veteranTypeDischarge = 'veteranTypeDischarge',
  veteranRank = 'veteranRank',
  veteranDd214 = 'veteranDd214',

  name = 'name',
  justice = 'justice',
  facility = 'facility',
  expectedReleasedAt = 'expectedReleasedAt',
  releasedCounty = 'releasedCounty',
  phone = 'phone',
  address = 'address',

  offenseCategory = 'offenseCategory',
  sexualOffenderRegistry = 'sexualOffenderRegistry',
  sbn = 'sbn',
  timeServed = 'timeServed',
  releasedAt = 'releasedAt',
  stateOrFederal = 'stateOrFederal',

  support = 'support',
  reentryServiceProvider = 'reentryServiceProvider',
  reentryPipeline = 'reentryPipeline',
  personalStrengths = 'personalStrengths',
  experienceSkills = 'experienceSkills',
  languages = 'languages',

  relativeExperience = 'relativeExperience',
  personalSummary = 'personalSummary',
  education = 'education',
  license = 'license',
  diversityInclusion = 'diversityInclusion',
  referredBy = 'referredBy',
}

export interface WizardData extends OffenseWizardType {
  veteranOrJustice: VeteranOrJustice[];

  firstName: string;
  lastName: string;
  phone?: string;
  // address
  address: string;
  city: string;
  state: State | string;
  postalCode: string;

  support: Support[];
  rspId?: string;
  isNewRspMember?: boolean;

  // diversityInclusion
  gender?: Gender;
  orientation?: Orientation;
  religion?: Religion;
  maritalStatus?: MaritalStatus;
  age?: Age;
  disability?: keyof typeof booleanOption;
  ethnicity?: Ethnicity;
  veteranStatus?: VeteranStatus;

  referredBy: ReferredBy;
  personalStrengths: PersonalStrengths[];
  experience: ExperienceSkills[];
  languages: Languages[];
  relativeExperience: (RelativeExperience & { stillWork: boolean })[];
  personalSummary: string;
  education: Education[];
  license: License[];

  // only justiceImpacted clients
  justiceStatus?: JusticeStatus;
  sexualOffenderRegistry?: keyof typeof booleanOption | null;
  sbn?: keyof typeof booleanOption | null;
  timeServed?: TimeServed;
  releasedAt?: ReleasedAt;
  stateOrFederal?: StateFederal;
  facility: string;
  expectedReleasedAt?: string;
  releasedCounty?: string;

  // only veteran clients
  veteranService?: VeteranBranchOfService[];
  veteranCampaigns?: VeteranCampaign[];
  veteranRank?: VeteranRank;
  veteranStartAtMonth?: string;
  veteranStartAtYear?: string;
  veteranEndAtMonth?: string;
  veteranEndAtYear?: string;
  veteranReservist?: keyof typeof booleanOption;
  veteranTypeDischarge?: VeteranTypeDischarge;
  veteranDd214?: keyof typeof booleanOption;
}

export const booleanOption = {
  yes: true,
  no: false,
};

type WizardFormFields<T> = T extends FieldType ? Omit<FieldComponentProps<T>, 'control'> : never;
export type CustomWizardField = (
  | WizardFormFields<FieldType>
  | (Omit<WizardChoicesProps, 'control'> & {
      type: 'choices';
    })
) & {
  conditionalProps?: (
    fields: Partial<
      WizardData & RelativeExperience & Education & License & Languages & { stillWork: boolean }
    >,
  ) => { [key: string]: any };
  conditionalPropsParams?: string[];
};

export type CustomWizardStep = Omit<WizardStep, 'fields' | 'id'> & {
  id: StepsClient;
  fields?: CustomWizardField[];
};

type FilterConditionally<Source, Condition> = Pick<
  Source,
  { [K in keyof Source]-?: Source[K] extends Condition ? K : never }[keyof Source]
>;

export type ArrayStepWizardData = FilterConditionally<WizardData, object[]>;
export type ArrayStepItem = ArrayStepWizardData[keyof ArrayStepWizardData][0];
export type ArrayStepItems = Array<ArrayStepItem>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type AvailableKeys = KeysOfUnion<ArrayStepItem>;
