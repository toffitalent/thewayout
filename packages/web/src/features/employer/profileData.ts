import { WizardStep } from '@disruptive-labs/ui';
import { Industry, NumberOfEmployers, YearsInBusiness } from '@two/shared';
import { states } from '@two/shared/src/constants/states';

export enum Steps {
  name = 'name',
  industry = 'industry',
  description = 'description',
  details = 'details',
  location = 'location',
}

export const stepsConfig: Record<Steps, { id: string; title: string }> = {
  [Steps.name]: { id: 'name', title: 'Employer Name' },
  [Steps.industry]: { id: 'industry', title: 'Industry' },
  [Steps.description]: { id: 'description', title: 'Company Description' },
  [Steps.details]: { id: 'details', title: 'Company Details' },
  [Steps.location]: { id: 'location', title: 'Company Location' },
};

export const steps: WizardStep[] = [
  {
    ...stepsConfig[Steps.name],
    fields: [
      {
        label: 'Name',
        name: 'name',
        type: 'text',
        placeholder: 'E.g. ABC Company',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    nextStep: 'industry',
  },
  {
    ...stepsConfig[Steps.industry],
    fields: [
      {
        label: 'Industry',
        name: 'industry',
        type: 'select',
        options: Object.values(Industry).map((el) => ({ label: el, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'description',
  },
  {
    ...stepsConfig[Steps.description],
    fields: [
      {
        label: 'Description',
        name: 'description',
        type: 'text',
        multiline: true,
        placeholder: 'Your Company description... ',
        validate: { required: true, maxLength: 1000 },
        helperText: 'Max 1,000 characters.',
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'details',
  },
  {
    ...stepsConfig[Steps.details],
    fields: [
      {
        label: 'Years in Business',
        name: 'yearsInBusiness',
        type: 'select',
        options: Object.values(YearsInBusiness).map((el) => ({ label: `${el} years`, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
      {
        label: 'Number of Employees',
        name: 'numberOfEmployees',
        type: 'select',
        options: Object.values(NumberOfEmployers).map((el) => ({ label: el, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'location',
  },
  {
    ...stepsConfig[Steps.location],
    fields: [
      {
        label: 'Address',
        name: 'address',
        type: 'text',
        placeholder: 'E.g. 459 S Hewitt St',
        validate: { required: true },
      },
      {
        label: 'City',
        name: 'city',
        type: 'text',
        placeholder: 'E.g. Los Angeles',
        validate: { required: true },
      },
      {
        label: 'State',
        name: 'state',
        type: 'select',
        options: Object.entries(states).map(([key, value]) => ({ value: key, label: value })),
        placeholder: 'Select',
        validate: { required: true },
      },
      {
        label: 'ZIP Code',
        name: 'postalCode',
        type: 'text',
        placeholder: 'E.g. 90255',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'confirm',
  },
];
