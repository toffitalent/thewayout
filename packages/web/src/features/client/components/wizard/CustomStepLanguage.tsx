import { WizardStep } from '@disruptive-labs/ui';
import { StepsClient } from '../../profileDataTypes';
import { ArrayStep } from './ArrayStep';

export const CustomStepLanguage = (props: Partial<WizardStep>) => {
  const { fields, nextStep } = props;
  if (!fields) {
    return null;
  }

  return (
    <ArrayStep
      minimalItems={0}
      id={StepsClient.languages}
      fields={fields}
      showListInEditMode={false}
      initialData={{ language: '', level: '' }}
      nextStep={nextStep}
      titleFieldName="language"
      text1FieldName="level"
    />
  );
};
