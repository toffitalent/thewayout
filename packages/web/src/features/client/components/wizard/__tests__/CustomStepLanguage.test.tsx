import { useWizardContext } from '@disruptive-labs/ui';
import userEvent from '@testing-library/user-event';
import { render, waitFor } from '@test';
import { steps } from '../../../profileData';
import { StepsClient } from '../../../profileDataTypes';
import { CustomStepLanguage } from '../CustomStepLanguage';

const languageStep = steps.find((el) => el.id === StepsClient.languages);

jest.mock('@disruptive-labs/ui', () => ({
  ...jest.requireActual('@disruptive-labs/ui'),
  useWizardContext: jest.fn(),
}));

describe('CustomStepLanguage', () => {
  beforeEach(() => {
    (useWizardContext as any).mockImplementation(() => ({
      data: {},
      setData: jest.fn(),
      goTo: jest.fn(),
    }));
  });

  test('renders correctly', async () => {
    const { container, findByText } = render(<CustomStepLanguage {...languageStep} />);
    await waitFor(async () => expect(await findByText('Language')).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

  test('skip step', async () => {
    const next = jest.fn();
    (useWizardContext as any).mockImplementation(() => ({
      data: {
        languages: [],
      },
      setData: jest.fn(),
      next,
    }));
    const { container, getByText } = render(<CustomStepLanguage {...languageStep} />);
    await userEvent.click(getByText('Skip'));
    expect(next).toBeCalled();
    expect(container).toMatchSnapshot();
  });

  test('press next & save button', async () => {
    const next = jest.fn();
    const setData = jest.fn();
    (useWizardContext as any).mockImplementation(() => ({
      data: {
        languages: [],
      },
      setData,
      next,
    }));
    const { container, getByText } = render(<CustomStepLanguage {...languageStep} />);
    const selectLang = container.querySelector('select[name=language]');
    if (selectLang) {
      userEvent.selectOptions(selectLang, 'english');
    }
    const selectLevel = container.querySelector('select[name=level]');
    if (selectLevel) {
      userEvent.selectOptions(selectLevel, 'elementary');
    }
    await waitFor(() => expect(getByText('Save & Next')).not.toBeDisabled());
    await userEvent.click(getByText('Save & Next'));
    expect(setData).toBeCalledWith({ languages: [{ language: 'english', level: 'elementary' }] });

    await userEvent.click(getByText('Save & Next'));
    expect(next).toBeCalled();
    expect(container).toMatchSnapshot();
  });
});
