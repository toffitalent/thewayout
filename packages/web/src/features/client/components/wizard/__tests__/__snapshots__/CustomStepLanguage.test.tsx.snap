// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomStepLanguage press next & save button 1`] = `
<div>
  <div
    class="maxWidth-96 marginRight-auto marginLeft-auto"
  >
    <div
      class="flexDirection-row justifyContent-center marginTop-8"
    >
      <button
        class="button md outline c-primary paddingRight-5 paddingLeft-5"
        type="button"
      >
        Add new
      </button>
      <button
        class="button md solid c-primary flex-1 marginLeft-5"
        type="button"
      >
        Save & Next
      </button>
    </div>
  </div>
</div>
`;

exports[`CustomStepLanguage renders correctly 1`] = `
<div>
  <div
    class="maxWidth-96 marginRight-auto marginLeft-auto"
  >
    <form>
      <div
        class="control"
        role="group"
      >
        <label
          class="label"
          for=":r0:"
        >
          Language
        </label>
        <div
          class="root primary md fluid md"
        >
          <select
            class="input"
            id=":r0:"
            name="language"
          >
            <option
              disabled=""
              hidden=""
              value=""
            >
              Select
            </option>
            <option
              value="english"
            >
              English
            </option>
            <option
              value="spanish"
            >
              Spanish
            </option>
            <option
              value="french"
            >
              French
            </option>
            <option
              value="chinese"
            >
              Chinese
            </option>
            <option
              value="german"
            >
              German
            </option>
            <option
              value="russian"
            >
              Russian
            </option>
            <option
              value="vietnamese"
            >
              Vietnamese
            </option>
            <option
              value="tagalog"
            >
              Tagalog
            </option>
            <option
              value="arabic"
            >
              Arabic
            </option>
            <option
              value="korean"
            >
              Korean
            </option>
          </select>
          <div
            class="icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,16c-0.3,0-0.5-0.1-0.7-0.3l-6-6c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l5.3,5.3l5.3-5.3c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-6,6C12.5,15.9,12.3,16,12,16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="control"
        role="group"
      >
        <label
          class="label"
          for=":r1:"
        >
          Level
        </label>
        <div
          class="root primary md fluid md"
        >
          <select
            class="input"
            id=":r1:"
            name="level"
          >
            <option
              disabled=""
              hidden=""
              value=""
            >
              Select
            </option>
            <option
              value="elementary"
            >
              Basic
            </option>
            <option
              value="basic"
            >
              Conversational
            </option>
            <option
              value="professional"
            >
              Proficient
            </option>
            <option
              value="nativeBilingual"
            >
              Fluent
            </option>
          </select>
          <div
            class="icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,16c-0.3,0-0.5-0.1-0.7-0.3l-6-6c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l5.3,5.3l5.3-5.3c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-6,6C12.5,15.9,12.3,16,12,16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="flexDirection-row justifyContent-center marginTop-8"
      >
        <button
          class="button md text c-primary paddingRight-5 paddingLeft-5"
          type="button"
        >
          Skip
        </button>
        <button
          class="button md solid c-primary disabled flex-1"
          disabled=""
          type="submit"
        >
          Save & Next
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`CustomStepLanguage skip step 1`] = `
<div>
  <div
    class="maxWidth-96 marginRight-auto marginLeft-auto"
  >
    <form>
      <div
        class="control"
        role="group"
      >
        <label
          class="label"
          for=":r2:"
        >
          Language
        </label>
        <div
          class="root primary md fluid md"
        >
          <select
            class="input"
            id=":r2:"
            name="language"
          >
            <option
              disabled=""
              hidden=""
              value=""
            >
              Select
            </option>
            <option
              value="english"
            >
              English
            </option>
            <option
              value="spanish"
            >
              Spanish
            </option>
            <option
              value="french"
            >
              French
            </option>
            <option
              value="chinese"
            >
              Chinese
            </option>
            <option
              value="german"
            >
              German
            </option>
            <option
              value="russian"
            >
              Russian
            </option>
            <option
              value="vietnamese"
            >
              Vietnamese
            </option>
            <option
              value="tagalog"
            >
              Tagalog
            </option>
            <option
              value="arabic"
            >
              Arabic
            </option>
            <option
              value="korean"
            >
              Korean
            </option>
          </select>
          <div
            class="icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,16c-0.3,0-0.5-0.1-0.7-0.3l-6-6c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l5.3,5.3l5.3-5.3c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-6,6C12.5,15.9,12.3,16,12,16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="control"
        role="group"
      >
        <label
          class="label"
          for=":r3:"
        >
          Level
        </label>
        <div
          class="root primary md fluid md"
        >
          <select
            class="input"
            id=":r3:"
            name="level"
          >
            <option
              disabled=""
              hidden=""
              value=""
            >
              Select
            </option>
            <option
              value="elementary"
            >
              Basic
            </option>
            <option
              value="basic"
            >
              Conversational
            </option>
            <option
              value="professional"
            >
              Proficient
            </option>
            <option
              value="nativeBilingual"
            >
              Fluent
            </option>
          </select>
          <div
            class="icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,16c-0.3,0-0.5-0.1-0.7-0.3l-6-6c-0.4-0.4-0.4-1,0-1.4s1-0.4,1.4,0l5.3,5.3l5.3-5.3c0.4-0.4,1-0.4,1.4,0s0.4,1,0,1.4l-6,6C12.5,15.9,12.3,16,12,16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="flexDirection-row justifyContent-center marginTop-8"
      >
        <button
          class="button md text c-primary paddingRight-5 paddingLeft-5"
          type="button"
        >
          Skip
        </button>
        <button
          class="button md solid c-primary disabled flex-1"
          disabled=""
          type="submit"
        >
          Save & Next
        </button>
      </div>
    </form>
  </div>
</div>
`;
